packages/lesswrong/lib/rateLimits/types.ts

```diff
@@ -29,39 +31,9 @@ export interface AutoRateLimit {
   itemsPerTimeframe: number, // number of items a user can post/comment/etc before triggering rate limit
   rateLimitType?: RateLimitType // short name used in analytics db
   rateLimitMessage: string // A message displayed to users when they are rate limited.
-
-  // The following parameters are optional, and if set, the rate limit will only apply to users who meet all thresholds:
-
-  karmaThreshold?: number, // if set, limit will only apply to users with karma less than the threshold
-  downvoteRatioThreshold?: number, // if set, limit will only apply to users who's ratio of received downvotes / total votes is higher than the listed threshold
-  last20KarmaThreshold?: number //  if set, limit only applies to users whose past 20 posts and comments karma total is less than N
-  last20PostKarmaThreshold?: number // if set, limit only applies to users whose past 20 post karma total is less than N
-  last20CommentKarmaThreshold?: number // if set, limit only applies to users whose past 20 comment karma total is less than N
-  downvoterCountThreshold?: number // if set, limit only applies to users whose past 20 posts and comments were downvoted by N or more people.
-  postDownvoterCountThreshold?: number // if set, limit only applies to users whose past 20 posts were downvoted by N or more people.
-  commentDownvoterCountThreshold?: number // if set, limit only applies to users whose past 20 comments were downvoted by N or more people.
-  lastMonthKarmaThreshold?: number // if set, limit only applies to votes on content from the last month
-  lastMonthDownvoterCountThreshold?: number // if set, limit only applies to users whose content in the last month was downvoted by N or more people.
+  isActive: (user: RateLimitUser, features?: AnyBecauseTodo) => boolean, 
```

Comment:
I think `features` should probably not actually be optional (nor any fields inside of it), since I think we're in all cases computing all the values we'd be checking against any of the thresholds a rate limit might have had.

(This next bit might feel a bit micro-manage-y, but the PR wasn't marked as a draft, and either way I'm confused by the development process implied by it.)  I'm surprised that it didn't seem worth defining the type for `features` before doing all the work to refactor the code to actually use it.  I think in cases like this it's nearly always a mistake to not do that first, not just because it'll catch random typos you make when accessing fields on the thing, but because it forces you to pull your working model out of head and into the code, which makes it much easier to notice mistakes/inconsistencies/confusions/etc.  As a random example, it would've surfaced the issue with optional chaining on the left side of a numeric comparison (which I was confused that the type checker didn't already catch for you, but then saw that `features` was optional - `any | undefined` gets simplified to `any`, as with any union type that includes an `any`).

Reply:
What about the universal rate limits that don't depend on the person?

Reply:
👁️ , thanks for the feedback! will type it out

Reply:
@jpaddison3 still seems fine to have the user be a required paramter, since there isn't obviously going to be a case where we're using this logic to try to rate limit activity by logged out users.  (There are a couple rate limits which don't depend on any details of the user already, or are always active; those just look like `isActive: () => true` or something.)